### ISSUE TITLE
index error in parafac with tensorflow backend

### ISSUE THREAD

**Issue by ischlag** [tensorly/tensorly#103]
With TensorFlow 1.12.0 as the backend I run into an index error when trying to use parafac. 

Try the following:
```import numpy as np
import tensorflow as tf
tf.__version__

tf.enable_eager_execution()
import tensorly as tl
tl.__version__

tl.set_backend("tensorflow")
T=tf.random_normal((8,8,8))

from tensorly.decomposition import parafac  
parafac(T, rank=1) 
````

My output:
```
In [1]: import numpy as np
In [2]: import tensorflow as tf
In [3]: tf.__version__                                                                                                             
Out[3]: '1.12.0'
In [4]: tf.enable_eager_execution()
In [5]: import tensorly as tl                                                                                                      
Using numpy backend.
In [6]: tl.__version__                                                                                                             
Out[6]: '0.4.3'
In [7]: tl.set_backend("tensorflow")                                                                                               
Using tensorflow backend.
In [8]: T=tf.random_normal((8,8,8))                                                                                                
2019-03-13 08:49:07.261789: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1432] Found device 0 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.7335
pciBusID: 0000:05:00.0
totalMemory: 7.92GiB freeMemory: 6.45GiB
2019-03-13 08:49:07.261821: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0
2019-03-13 08:49:07.508069: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-03-13 08:49:07.508105: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0 
2019-03-13 08:49:07.508112: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N 
2019-03-13 08:49:07.508281: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6211 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080, pci bus id: 0000:05:00.0, compute capability: 6.1)

In [9]: from tensorly.decomposition import parafac  
In [10]: parafac(T, rank=1)                                                                                                        
2019-03-13 08:49:24.971196: I tensorflow/core/kernels/cuda_solvers.cc:159] Creating CudaSolver handles for stream 0x661c270
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-10-a4f1e5f3d795> in <module>
----> 1 parafac(T, rank=1)

~/.virtualenvs/py3/lib/python3.6/site-packages/tensorly/decomposition/candecomp_parafac.py in parafac(tensor, rank, n_iter_max, init, svd, tol, orthogonalise, random_state, verbose, return_errors)
    177                 if i != mode:
    178                     pseudo_inverse = pseudo_inverse*tl.dot(tl.transpose(factor), factor)
--> 179             factor = tl.dot(unfold(tensor, mode), khatri_rao(factors, skip_matrix=mode))
    180             factor = tl.transpose(tl.solve(tl.transpose(pseudo_inverse), tl.transpose(factor)))
    181             factors[mode] = factor

~/.virtualenvs/py3/lib/python3.6/site-packages/tensorly/tenalg/_khatri_rao.py in khatri_rao(matrices, skip_matrix, reverse)
     68         return matrices[0]
     69 
---> 70     n_columns = matrices[0].shape[1]
     71 
     72     # Optional part, testing whether the matrices have the proper size

~/.virtualenvs/py3/lib/python3.6/site-packages/tensorflow/python/framework/tensor_shape.py in __getitem__(self, key)
    614         return TensorShape(self._dims[key])
    615       else:
--> 616         return self._dims[key]
    617     else:
    618       if isinstance(key, slice):

IndexError: list index out of range
```

**Comment by ischlag** [comment]
I'm puzzled by this because the following works.
```
import numpy as np
import tensorflow as tf
tf.enable_eager_execution()
import tensorly as tl
tl.set_backend("tensorflow")

T = tf.random_normal((8,8,8))
from tensorly.decomposition.candecomp_parafac import initialize_factors 
factors = initialize_factors(T, rank=1)
```

**Comment by ischlag** [comment]
I found the solution to this bug.

The problem is that on line 179 in candecomp_parafac.py  
`tl.solve(tl.transpose(pseudo_inverse), tl.transpose(factor))` with a rank of 1 will return (n,) shape and not (n,1), hence the index error trying to lookup the second index later on.

The problem is that tf.solve will not return a 2 dimensional array if your factor "matrix" is only a single column.

I fixed this locally by simply doing the following before adding the solve result to the list of factors.
```
            factor = tl.transpose(tl.solve(tl.transpose(pseudo_inverse), tl.transpose(factor)))
            if factor.ndim == 1:
                factor = tl.reshape(factor, (-1, 1))
            factors[mode] = factor
```
